File: flutter/lib/core/services/chat_service.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import '../../shared/models/chat_model.dart';

/// Chat Service Provider
final chatServiceProvider = Provider<ChatService>((ref) {
  return ChatService();
});

/// Chat Service
/// 
/// Handles all chat-related operations.
class ChatService {
  final SupabaseClient _client = Supabase.instance.client;

  /// Get chat pricing configuration
  Future<ChatPricingModel> getChatPricing() async {
    try {
      final response = await _client
          .from('chat_pricing')
          .select()
          .eq('is_active', true)
          .maybeSingle();

      if (response == null) {
        return const ChatPricingModel();
      }

      return ChatPricingModel(
        ratePerMinute: (response['rate_per_minute'] as num?)?.toDouble() ?? 2,
        womenEarningRate: (response['women_earning_rate'] as num?)?.toDouble() ?? 2,
        videoRatePerMinute: (response['video_rate_per_minute'] as num?)?.toDouble() ?? 10,
        videoWomenEarningRate: (response['video_women_earning_rate'] as num?)?.toDouble() ?? 5,
        minWithdrawalBalance: (response['min_withdrawal_balance'] as num?)?.toDouble() ?? 10000,
        currency: response['currency'] as String? ?? 'INR',
      );
    } catch (e) {
      return const ChatPricingModel();
    }
  }

  /// Get chat messages
  Future<List<ChatMessageModel>> getChatMessages(String chatId, {int limit = 100}) async {
    try {
      final response = await _client
          .from('chat_messages')
          .select()
          .eq('chat_id', chatId)
          .order('created_at', ascending: true)
          .limit(limit);

      return (response as List)
          .map((json) => ChatMessageModel(
                id: json['id'],
                chatId: json['chat_id'],
                senderId: json['sender_id'],
                receiverId: json['receiver_id'],
                message: json['message'],
                translatedMessage: json['translated_message'],
                isTranslated: json['is_translated'] ?? false,
                isRead: json['is_read'] ?? false,
                flagged: json['flagged'] ?? false,
                flagReason: json['flag_reason'],
                createdAt: json['created_at'] != null
                    ? DateTime.parse(json['created_at'])
                    : null,
              ))
          .toList();
    } catch (e) {
      return [];
    }
  }

  /// Send a message
  Future<ChatMessageModel?> sendMessage({
    required String chatId,
    required String senderId,
    required String receiverId,
    required String message,
  }) async {
    try {
      final response = await _client
          .from('chat_messages')
          .insert({
            'chat_id': chatId,
            'sender_id': senderId,
            'receiver_id': receiverId,
            'message': message,
          })
          .select()
          .single();

      return ChatMessageModel(
        id: response['id'],
        chatId: response['chat_id'],
        senderId: response['sender_id'],
        receiverId: response['receiver_id'],
        message: response['message'],
        createdAt: DateTime.parse(response['created_at']),
      );
    } catch (e) {
      return null;
    }
  }

  /// Get active chat sessions for user
  Future<List<ChatSessionModel>> getActiveSessions(String userId) async {
    try {
      final response = await _client
          .from('active_chat_sessions')
          .select()
          .or('man_user_id.eq.$userId,woman_user_id.eq.$userId')
          .eq('status', 'active');

      return (response as List)
          .map((json) => ChatSessionModel(
                id: json['id'],
                chatId: json['chat_id'],
                manUserId: json['man_user_id'],
                womanUserId: json['woman_user_id'],
                status: json['status'],
                startedAt: json['started_at'] != null
                    ? DateTime.parse(json['started_at'])
                    : null,
                endedAt: json['ended_at'] != null
                    ? DateTime.parse(json['ended_at'])
                    : null,
                totalMinutes: (json['total_minutes'] as num?)?.toDouble() ?? 0,
                totalEarned: (json['total_earned'] as num?)?.toDouble() ?? 0,
                ratePerMinute: (json['rate_per_minute'] as num?)?.toDouble() ?? 2,
                lastActivityAt: json['last_activity_at'] != null
                    ? DateTime.parse(json['last_activity_at'])
                    : null,
              ))
          .toList();
    } catch (e) {
      return [];
    }
  }

  /// Mark messages as read
  Future<void> markMessagesAsRead(String chatId, String receiverId) async {
    try {
      await _client
          .from('chat_messages')
          .update({'is_read': true})
          .eq('chat_id', chatId)
          .eq('receiver_id', receiverId)
          .eq('is_read', false);
    } catch (e) {
      // Handle error silently
    }
  }

  /// Subscribe to new messages
  RealtimeChannel subscribeToMessages(
    String chatId,
    void Function(ChatMessageModel message) onMessage,
  ) {
    return _client.channel('chat:$chatId').onPostgresChanges(
      event: PostgresChangeEvent.insert,
      schema: 'public',
      table: 'chat_messages',
      filter: PostgresChangeFilter(
        type: PostgresChangeFilterType.eq,
        column: 'chat_id',
        value: chatId,
      ),
      callback: (payload) {
        final json = payload.newRecord;
        onMessage(ChatMessageModel(
          id: json['id'],
          chatId: json['chat_id'],
          senderId: json['sender_id'],
          receiverId: json['receiver_id'],
          message: json['message'],
          translatedMessage: json['translated_message'],
          isTranslated: json['is_translated'] ?? false,
          isRead: json['is_read'] ?? false,
          createdAt: json['created_at'] != null
              ? DateTime.parse(json['created_at'])
              : null,
        ));
      },
    ).subscribe();
  }

  /// Process chat billing (via database function)
  Future<Map<String, dynamic>> processChatBilling(
    String sessionId,
    double minutes,
  ) async {
    try {
      final response = await _client.rpc('process_chat_billing', params: {
        'p_session_id': sessionId,
        'p_minutes': minutes,
      });

      return response as Map<String, dynamic>;
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }

  /// End chat session
  Future<bool> endChatSession(String sessionId, {String? reason}) async {
    try {
      await _client.from('active_chat_sessions').update({
        'status': 'ended',
        'ended_at': DateTime.now().toIso8601String(),
        'end_reason': reason,
      }).eq('id', sessionId);

      return true;
    } catch (e) {
      return false;
    }
  }
}

File: src/services/chat.service.ts
/**
 * Chat Service
 * 
 * Handles all chat-related API calls.
 */

import { supabase } from '@/integrations/supabase/client';

export interface ChatMessage {
  id: string;
  chat_id: string;
  sender_id: string;
  receiver_id: string;
  message: string;
  translated_message: string | null;
  is_translated: boolean | null;
  is_read: boolean | null;
  created_at: string;
}

export interface ChatSession {
  id: string;
  chat_id: string;
  man_user_id: string;
  woman_user_id: string;
  status: string;
  started_at: string;
  ended_at: string | null;
  total_minutes: number;
  total_earned: number;
  rate_per_minute: number;
}

export interface ChatPricing {
  ratePerMinute: number;
  womenEarningRate: number;
  videoRatePerMinute: number;
  videoWomenEarningRate: number;
  minWithdrawalBalance: number;
  currency: string;
}

const DEFAULT_PRICING: ChatPricing = {
  ratePerMinute: 4,              // Men pay ₹4/min for chat (matches DB)
  womenEarningRate: 2,           // Indian women earn ₹2/min for chat (admin configurable)
  videoRatePerMinute: 8,         // Men pay ₹8/min for video
  videoWomenEarningRate: 4,      // Women earn ₹4/min for video
  minWithdrawalBalance: 5000,
  currency: 'INR',
};
// Note: Non-Indian women earn ₹0/min - checked via is_earning_eligible flag in ChatEarningsDisplay

/**
 * Get chat pricing configuration
 */
export async function getChatPricing(): Promise<ChatPricing> {
  const { data, error } = await supabase
    .from('chat_pricing')
    .select('*')
    .eq('is_active', true)
    .maybeSingle();

  if (error || !data) {
    return DEFAULT_PRICING;
  }

  return {
    ratePerMinute: data.rate_per_minute || DEFAULT_PRICING.ratePerMinute,
    womenEarningRate: data.women_earning_rate || DEFAULT_PRICING.womenEarningRate,
    videoRatePerMinute: data.video_rate_per_minute || DEFAULT_PRICING.videoRatePerMinute,
    videoWomenEarningRate: data.video_women_earning_rate || DEFAULT_PRICING.videoWomenEarningRate,
    minWithdrawalBalance: data.min_withdrawal_balance || DEFAULT_PRICING.minWithdrawalBalance,
    currency: data.currency || DEFAULT_PRICING.currency,
  };
}

/**
 * Get chat messages for a session
 */
export async function getChatMessages(
  chatId: string,
  limit = 100
): Promise<ChatMessage[]> {
  const { data } = await supabase
    .from('chat_messages')
    .select('*')
    .eq('chat_id', chatId)
    .order('created_at', { ascending: true })
    .limit(limit);

  return data || [];
}

/**
 * Send a chat message
 */
export async function sendMessage(
  chatId: string,
  senderId: string,
  receiverId: string,
  message: string
): Promise<{ success: boolean; messageId?: string; error?: string }> {
  // Content moderation - block phone numbers, emails, social media
  const { moderateMessage } = await import('@/lib/content-moderation');
  const moderationResult = moderateMessage(message);
  if (moderationResult.isBlocked) {
    return { success: false, error: moderationResult.reason || 'This message contains prohibited content.' };
  }

  const { data, error } = await supabase
    .from('chat_messages')
    .insert({
      chat_id: chatId,
      sender_id: senderId,
      receiver_id: receiverId,
      message,
    })
    .select('id')
    .single();

  if (error) {
    return { success: false, error: error.message };
  }

  return { success: true, messageId: data.id };
}

/**
 * Get active chat sessions for a user
 */
export async function getActiveChatSessions(userId: string): Promise<ChatSession[]> {
  const { data } = await supabase
    .from('active_chat_sessions')
    .select('*')
    .or(`man_user_id.eq.${userId},woman_user_id.eq.${userId}`)
    .eq('status', 'active');

  return data || [];
}

/**
 * Mark messages as read
 */
export async function markMessagesAsRead(
  chatId: string,
  receiverId: string
): Promise<void> {
  await supabase
    .from('chat_messages')
    .update({ is_read: true })
    .eq('chat_id', chatId)
    .eq('receiver_id', receiverId)
    .eq('is_read', false);
}

/**
 * Process chat billing (uses database function)
 */
export async function processChatBilling(
  sessionId: string,
  minutes: number
): Promise<{ success: boolean; error?: string }> {
  const { data, error } = await supabase.rpc('process_chat_billing', {
    p_session_id: sessionId,
    p_minutes: minutes,
  });

  if (error) {
    return { success: false, error: error.message };
  }

  const result = data as unknown as { success: boolean; error?: string };
  return result;
}

/**
 * End chat session (synced with Flutter)
 */
export async function endChatSession(
  sessionId: string,
  reason?: string
): Promise<boolean> {
  const { error } = await supabase
    .from('active_chat_sessions')
    .update({
      status: 'ended',
      ended_at: new Date().toISOString(),
      end_reason: reason || null,
    })
    .eq('id', sessionId);

  return !error;
}

/**
 * Subscribe to new messages (real-time)
 */
export function subscribeToMessages(
  chatId: string,
  onMessage: (message: ChatMessage) => void
) {
  return supabase
    .channel(`chat:${chatId}`)
    .on(
      'postgres_changes',
      {
        event: 'INSERT',
        schema: 'public',
        table: 'chat_messages',
        filter: `chat_id=eq.${chatId}`,
      },
      (payload) => {
        onMessage(payload.new as ChatMessage);
      }
    )
    .subscribe();
}
